onnxruntime/contrib_ops/cuda/transformers/generation_cuda_impl.cu

```diff
@@ -0,0 +1,775 @@
+                                                                 d_sorted_indices,
+                                                                 d_logits_in_out,
+                                                                 top_p,
+                                                                 filter_value,
+                                                                 batch_size,
+                                                                 vocab_size);
+  } else {
+    FilterLogitsKernel<<<gridSize, blockSize, 0, stream>>>(d_sorted_logits_in,
+                                                           d_sorted_indices,
+                                                           d_logits_in_out,
+                                                           1 - top_p,
+                                                           filter_value,
+                                                           min_tokens_to_keep,
+                                                           batch_size,
+                                                           vocab_size);
+  }
+}
+
+template void LaunchFilterLogitsKernel(float* d_sorted_logits_in,
+                                       const int* d_sorted_indices,
+                                       float* d_logits_in_out,
+                                       float top_p,
+                                       float filter_value,
+                                       int min_tokens_to_keep,
+                                       int batch_size,
+                                       int vocab_size,
+                                       cudaStream_t stream,
+                                       bool is_descending);
+
+template void LaunchFilterLogitsKernel(float* d_sorted_logits_in,
+                                       const int* d_sorted_indices,
+                                       half* d_logits_in_out,
+                                       float top_p,
+                                       float filter_value,
+                                       int min_tokens_to_keep,
+                                       int batch_size,
+                                       int vocab_size,
+                                       cudaStream_t stream,
+                                       bool is_descending);
+
+
+// Ref: https://github.com/pytorch/pytorch/blob/release/1.13/aten/src/ATen/native/cuda/MultinomialKernel.cu
+template <typename scalar_t, typename accscalar_t, unsigned TPB>
+__global__ void sampleMultinomialOnce(int64_t* dest,
+                                      int distributions,
+                                      int categories,
+                                      scalar_t* sampled,
+                                      scalar_t* dist,
+                                      int stride_dist, // dist->stride(0)
+                                      int stride_categories, // dist->stride(1)
```

Comment:
>stride_categories [](http://example.com/codeflow?start=42&length=17)

I don't understand why stride_categories is batch_size instead of 1?

Reply:
good catch! it should be one